### PR TITLE
[FLINK-23381][state] Back-pressure on reaching state change to upload limit

### PR DESCRIPTION
**Pull request by rkhachatryan**
## What is the purpose of the change

Propagate back-pressure from changelog storage to (stream) task (record processing and checkpoionting).
The 1st commit replaces failing with blocking caller on reaching the limit.
The 2nd commit adds availability provider to StreamTask, so it shouldn't block in most cases. Checkpointing will still block.
The change wasn't benchmarked yet.

## Verifying this change
  - `BatchingStateChangeUploaderTest.testBackPressure`
  - `BatchingStateChangeUploaderTest.testInterruptedWhenBackPressured`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): yes
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? no

### CODE REVIEW

**Comment by complone** [comment]
Hello, can you explain how the back pressure draft added in the flink1.15 version and one of the state backends specified in the org/apache/flink/state/changelog/ChangelogStateBackendTestUtils.java class registered monitoring indicators in EmbeddedRocksDBStateBackend.java Whether the methods are consistent, the registered MetricGroup refers to the indicators mentioned in the draft here. The location of the code that needs to be recorded, I read it according to the logic of the code, the flow chart is as follows：

https://www.processon.com/view/link/6110a74d0e3e741f86d21e11

My question is whether my work is as described above for [FLINK-23486](https://issues.apache.org/jira/browse/FLINK-23486)

**Comment by rkhachatryan** [comment]
Sorry @complone but don't fully understand the question.
This PR is not directly related to metrics.
I also do not understand the purpose of the flowchart, but it also seems unrelated to this PR.

**Comment by curcur** [comment]
Thanks @rkhachatryan for the PR, I have a few high-level + implementation questions. Let's discuss offline:

- Item 1
seizeCapacity: multiple task thread?

1. hasCapacity(): inFlightBytesCounter < maxBytesInFlight;
   isAvaialble(): what is that, how it is different from hasCapacity()
   
2. Even waken up, it still not guranteed to have enough capacity
	(inFlightBytesCounter + bytes) still possible to exceed capacity
	what's the purpose to split seizeCapacity into these two steps?
	
- Item 2
releaseCapacity: IO tread, single IO thread, multiple IO thread?
How it is related to upload task scheduler?

UploadThrottle - AvailabilityProvider
BatchingStateChangeUploader contains AvailabilityProvider

- Item 3
BatchingStateChangeUploader.upload
even if it grabs the lock and seize capacity, it may still not able to upload all bytes
Why it is designed this way

(Let's discuss this offline)

- Item 4
BatchingStateChangeUploader per FsStateChangelogStorage inlcuded in TaskStateManager
per TM

- Item 5
Why it is allowed to be null?
environment.getTaskStateManager().getStateChangelogStorage() == null

- Item 6
Now Backpressure seems has one more state:
idleTimeMsPerSecond
busyTimeMsPerSecond
backPressuredTimeMsPerSecond -- backpressure because of downstream / backpressure because of changelog IO

- Item 7
I found there are several locks in a few different places, are they all necessary (asking this because of the DFS PR is not reviewed by me, so needs a bit more input)

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java
Comment (pnowojski) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java:58]: If we want to go with this backpressured metric this code should be probably wrapped with
```
backPressuredTimeMsPerSecond.markStart();
(...)
backPressuredTimeMsPerSecond.markEnd();
```

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java
Comment (pnowojski) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java:74]: complete it outside of the lock?

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java
Comment (pnowojski) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java:31]: Per our offline discussion, maybe this could be made thread unsafe, and access to this could be synchronised/protected from `BatchingStateChangeUploader`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:545]: For the sake of simplicity, in this PR and in the MVP maybe it would be best to account time when changelog/state backend is not available as `busy`? This would be easier to implement as it would naturally account consistently waiting on the availability future and on `lock.wait();`.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java
Comment (rkhachatryan) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java:74]: Good point!

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java
Comment (rkhachatryan) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java:58]: As discussed offline, I'll remove the use of `backPressuredTimePerSecond` in this PR, so any waiting on changelog will be reported as busy (which is consistent with other state backends currently).

The mertric then should be addressed in FLINK-24402.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:399]: After thinking a bit more about `getStateBackendAvailabilityProvider`, it doesn't seem the right API.

Inside a task, there can be multiple operators and therefore multiple state backends.
In general (without sharing the underlying writers), each can have its own availability provider.
So such API (`getTaskStateManager().getStateBackendAvailabilityProvider()`) wouldn't make sense.

WDYT?

**Comment by curcur** [comment]
Hey folks, I’ve spent some time reading through the `FsStateChangelogWriter` and related `DFS` uploading implementation. Now I have a bit better understanding but with more questions (We can discuss offline if necessary).

Before posting my questions here, let me briefly describe my understanding (This is needed because it relates to my question).

Please correct me if I misunderstand

1. `StateChangeUploader` is per TM, shared amongst all tasks belonging to the same job (same TM)
2. For each `ChangelogKeyedStateBackend`, it has a separate writer (StateChangelogWriter), maintaining `activeChangeSet`(in memory), to which the writer writes to
3. persist (`persistInternal`) happens in two cases: 1). enough data in memory (`preEmptivePersistThresholdInBytes`); 2). checkpointing
4. What persist does is to put an UploadTask to `StateChangeUploader` 's upload queue, waiting to schedule to upload
5. The above all occur in the task thread
6. Then scheduling thread (SchedulerFactory.create(1, "ChangelogUploadScheduler", LOG)) schedule task uploading (take UploadTasks away from StateChangeUploader 's upload queue
7. IO thread (retryingExecutor.execute(retryPolicy, () -> delegate.upload(tasks)))  execute uploading tasks

=====
**Questions (about availability provider)**

1. The UploadThrottle (availability provider)  is shared amongst all tasks belonging to the same job; each time to check the availability, we have to grab the lock, whether it has a performance issue. From what I can see, the current `RecordWriter` and `InputProcessor` is per task and without lock; whether we should consider implementing changelog availablity provider per task as well?
2. for check `changelogWriterAvailabilityProvider.isApproximatelyAvailable()`, when checking `getAvailableFuture() == AVAILABLE`, it is outside the lock, so when doing the check, the available future can be complete?
3. Even if the available future is available, it means some task is able to proceed updating states, not necessary all tasks (other tasks might still be blocked on waiting)

**Questions (about the locks)**
I found there are three locks in three different classes (not sure whether there are more).
`UploadThrottle`
`BatchingStateChangeUploader`
`FsStateChangelogWriter`

But from the description above (the paragraph of "Please correct me if I misunderstand"), there is only one critical piece, that is "StateChangeUploader 's upload queue"

1. Several writers need to put uploadtasks into the queue
2. Scheduler need to schedule the tasks from the queue
3. AvaiablityProvider needs to check availability of the queue
So, from what I can see, these three locks can be merged into one? Is there any reason to separate them?
As discussed, more locks are easier to cause deadlock and make the code more difficult to reason about.

**Questions (about the backpressure metric)**
They are not mutually exclusive: but we can postpone this discussion when designing metrics.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java
Comment (dawidwys) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java:138]: did you mean `interrupted()` here?

Comment (rkhachatryan) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java:138]: No, I actually mean `interrupt()` to restore interrupt status (since `InterruptedException` is caught and wrappen, and caller might not check the cause but might check the status).

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:399]: Good point. I don't know how to properly solve it.

+0 for the current proposal, maybe we can revisit it in the future if someone comes up with a better alternative? In the future we could also rename it to `getTaskStateManager().getAvailabilityProvider()`, and how it's implemented underneath would be an implementation detail from the `StreamTask`'s perspective. But for now a more descriptive `getTaskStateManager().getStateChangelogStorage().getAvailabilityProvider()` maybe is a good enough solution.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java
Comment (pnowojski) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java:31]: Wouldn't this be solved by simplifying this class to a dumb accounting class of reserved/seized bytes?

So in `BatchingStateChangeUploader` instead of current calls:
```
uploadThrottle.seizeCapacity(size);
synchronized (scheduled) {
   foo();
}
```
```
(...)
uploadThrottle.releaseCapacity(preComputedTaskSize);
(...)
```
We would have:
```
synchronized (scheduled) {
   while (!uploadThrottle.seizeCapacity(size)) { //return true if successfully seized
     scheduled.wait();
   }
   foo();
}
```
```
(...)
CompletableFuture<?> toNotify = null;
synchronized (scheduled) {
  if (uploadThrottle.releaseCapacity(preComputedTaskSize)) { // return true if released enough to unblock
    toNotify = availabilityHelper.getUnavailableToResetAvailable();
  }
  scheduled.notifyAll();
}
if (toNotify != null) {
  toNotify.complete(null);
}
(...)
```

Comment (rkhachatryan) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java]: I don't have a strong opinion because moving that much functionality from `UploadThrottle` to `BatchingStateChangeUploader` complicates the latter quite a bit
so I'll change the code as you're suggesting.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:399]: I agree, also in the future there could be other use cases for changelog, so it makes sense to keep it as is for now.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:299]: These are the top 10 regressions when comparing [this PR](http://codespeed.dak8s.net:8080/job/flink-benchmark-request/553/artifact/jmh-result.csv/*view*/) and [master](http://codespeed.dak8s.net:8080/job/flink-master-benchmarks/8589/artifact/jmh-result.csv/*view*/):

MultiInputCheckpointingTimeBenchmark.checkpointMultiInput | 2.57734 | 89.77%
-- | -- | --
SerializationFrameworkMiniBenchmarks.serializerPojo | 629.770475 | 91.29%
StreamNetworkThroughputBenchmarkExecutor.networkThroughput | 36497.36957 | 93.05%
BlockingPartitionBenchmark.compressedFilePartition | 8733.624685 | 94.31%
WindowBenchmarks.globalWindow | 6870.474052 | 96.32%
MultipleInputBenchmark.multiInputMapSink | 18164.0954 | 97.22%
SerializationFrameworkMiniBenchmarks.serializerRow | 848.132961 | 97.31%
TwoInputBenchmark.twoInputMapSink | 17246.67162 | 97.93%
WindowBenchmarks.sessionWindow | 820.272928 | 98.18%
InputBenchmark.mapSink | 30215.03625 | 98.28%

All these benchmarks seem to be noisy, so I think the performance is fine, WDYT?

(I ran with and without `null` optimization without any significant difference; these results are without `null`, i.e. lambda).

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:299]: Which version/parameters did you mean when reporting networkThroughput? I presume `1000,1ms`? If so, I think those results are fine. Those results are well within the noise of those benchmarks. I think except of `WindowBenchmarks.globalWindow`, all of the other results are within the min/max of last 10 runs.

Nevertheless we should check the long term trends after merging this PR.

Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:299]: Yes, those `networkThroughput` results are for `1000,1ms`.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java
Comment (pnowojski) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java:259]: Why indirect `Runnable`? Can not we call `releaseCapacity(size)` directly here?

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java
Comment (pnowojski) [inline flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java:276]: Can you include here in a comment the explanation why we think it's fine to have this unsynchronised, despite that other task threads might actually make this unavailable?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:299]: nit: Re the `null` version. I think I would sleep better with it here. Maybe the change is barely visible (1% degradation?) or even if it's not visible, it would be a one less potential issue to wonder/investigate/optimise for someone in the future.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:488]: Why `isApproximatelyAvailable` and not `isAvailable`? Isn't this actually an anti optimisation? If the `changelogWriterAvailabilityProvider.getAvailableFuture() != AVAILABLE`, it should be actually cheaper to check `changelogWriterAvailabilityProvider.getAvailableFuture().isDone()` rather then suspend mailbox and wake it up via `CompletableFuture#thenRun` callback.

**Review by pnowojski — COMMENTED** [review]
Thanks for the update @rkhachatryan . The change looks now mostly good to me. I've left a couple of questions/minor comments.

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java
Comment (rkhachatryan) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeUploader.java:259]: My intention was to make it clear what is captured in terms of references at runtime;
but apparrently it doesn't make it more clear, so I'll change back to a direct call and make `wrapWithSizeUpdate` non-static.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java
Comment (rkhachatryan) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:488]: You're right, I'll change to `isAvailable`.

**Review by pnowojski — APPROVED** [review]
Thanks for the changes

LGTM % squashing and assuming green azure and that @curcur do not have any further comments?

**Comment by curcur** [comment]
Thanks! I do not have further concerns! Please go ahead!

File: flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java
Comment (curcur) [reply flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/UploadThrottle.java]: here `waiting` means blocking-waiting.
- blocking waiting is reported as busy
- non-blocking waiting is reported ?? (follow-up)

**Pull request merged**
